sdk/test-utils/recorder-new/src/core-v2-recorder.ts

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { WebResourceLike } from "@azure/core-http";
+import {
+  createDefaultHttpClient,
+  createPipelineRequest,
+  HttpClient,
+  PipelinePolicy,
+  PipelineRequest,
+  PipelineResponse,
+  SendRequest
+} from "@azure/core-rest-pipeline";
+
+const paths = {
+  playback: "/playback",
+  record: "/record",
+  start: "/start",
+  stop: "/stop"
+};
+
+/**
+ * Helper class to manage the recording state to make sure the proxy-tool is not flooded with unintended requests.
+ *
+ * => then start
+ * => run the runAsync
+ * => stop record
+ * => start playback
+ * => stop playback
+ */
+export class RecordingStateManager {
+  public state: "started" | "stopped" | undefined;
+
+  /**
+   * validateState
+   */
+  public validateState(currentFlow: "starting" | "stopping") {
+    if (currentFlow === "starting") {
+      if (this.state === "started") {
+        throw new Error("Recorder Error: Already started, should not have called start again.");
+      }
+    }
+    if (currentFlow === "stopping") {
+      if (this.state === "stopped") {
+        throw new Error("Recorder Error: Already stopped, should not have called stop again.");
+      }
+      if (this.state !== "started") {
+        throw new Error("Recorder Error: Please start before calling stop.");
+      }
+    }
+    if (currentFlow === "starting") {
+      if (this.state !== "stopped" && this.state !== undefined) {
+        throw new Error("Recorder Error: Please stop before calling start.");
+      }
+    }
+    if (currentFlow === "stopping") {
+      if (this.state !== "started") {
+        throw new Error("Recorder Error: Please start before calling stop.");
+      }
+    }
+  }
+
+  /**
+   * setState
+   */
+  public setState(state: "started" | "stopped") {
+    this.state = state;
+  }
+}
+
+export class TestProxyHttpClient {
```

Comment:
On second thought, I only see a general test, I don’t see unit tests. While I appreciate the high level test, because it shows things will work rather transparently, the logic in this project is fairly convoluted. Please add unit tests for all of the expected behaviors of this class.

Reply:
I also had the feeling this code came about based on other languages, but it seems like this is first being developed in JS. I believe we should be very careful with the design if JS is going to be the source of truth for these clients. Let’s showcase everything we’re expecting this source code to do through individual unit tests.